app/src/main/java/com/babylon/wallet/android/presentation/settings/securitycenter/SecurityCenterViewModel.kt

```diff
@@ -1,43 +1,54 @@
 package com.babylon.wallet.android.presentation.settings.securitycenter
 
 import androidx.lifecycle.viewModelScope
+import com.babylon.wallet.android.di.coroutines.DefaultDispatcher
 import com.babylon.wallet.android.domain.model.SecurityProblem
 import com.babylon.wallet.android.domain.usecases.GetSecurityProblemsUseCase
 import com.babylon.wallet.android.presentation.common.StateViewModel
 import com.babylon.wallet.android.presentation.common.UiState
 import dagger.hilt.android.lifecycle.HiltViewModel
-import kotlinx.coroutines.flow.update
-import kotlinx.coroutines.launch
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
 import javax.inject.Inject
 
 @HiltViewModel
 class SecurityCenterViewModel @Inject constructor(
-    private val getSecurityProblemsUseCase: GetSecurityProblemsUseCase
+    getSecurityProblemsUseCase: GetSecurityProblemsUseCase,
+    @DefaultDispatcher dispatcher: CoroutineDispatcher
 ) : StateViewModel<SecurityCenterViewModel.SecurityCenterUiState>() {
 
     override fun initialState(): SecurityCenterUiState {
-        return SecurityCenterUiState()
+        return SecurityCenterUiState.Loading
     }
 
     init {
-        viewModelScope.launch {
-            getSecurityProblemsUseCase().collect { problems ->
-                _state.update { state ->
-                    state.copy(
-                        securityProblems = problems
+        getSecurityProblemsUseCase()
+            .onEach { securityProblems ->
+                _state.emit(
+                    SecurityCenterUiState.Data(
+                        securityProblems = securityProblems
                     )
-                }
+                )
             }
-        }
+            .flowOn(dispatcher)
```

Comment:
do you think we should better move the `.flowOn(dispatcher)` in `GetSecurityProblemsUseCase` and maybe to add it to `GetEntitiesWithSecurityPromptUseCase` too?

Reply:
Also I don't remember where I read it but rendering UI (_state.emit) should be running on main thread.

Reply:
But then the computation done in `viewModel` would be performed on the main thread. Maybe it's not the case here as there isn't much computation, but in other cases we should consider using the dispatcher within the `viewModel` as well. What's your concern about using it within the `viewModel`?

Reply:
> Also I don't remember where I read it but rendering UI (_state.emit) should be running on main thread.

I'm not so sure about this, can't see a reason for it, do you have any references?

Reply:
> I'm not so sure about this, can't see a reason for it, do you have any references? 

I cant recall where I read about it, but as far as I remember it was a design decision. 
Anyway it's not a hard requirement, and if I find the reference I will post it

Reply:
All the computation done upon collecting or in the operators called from the `viewModel` will still be on main thread if moving the `flowOn` to the use cases. The only computation done in this case if moving the `flowOn` to the `GetSecurityProblemsUseCase` is the initialization of the 3 boolean values inside the `UiState` and it won't make a difference here, but in other cases it will definitely do, as there are also calls to Sargon, many list iterations, etc (`WalletViewModel` is a good example for this). That's why I wondered what's the reason against having the `flowOn(dispatcher)` within the `viewModel`, because in this case I agree that it won't make a difference at all, while in others it will.

Reply:
Now that I think of it, the only reason I can see for the state to not be updated from the background thread is to prevent multiple modifications from different threads (concurrent modifications -> it won't even crash, but it might lead to a bad state), but this shouldn't be the case if following the [unidirectional data flow](https://developer.android.com/develop/ui/compose/architecture#udf) pattern.

Reply:
Here is a snippet from the StateFlow docs:

<img width="653" alt="Screenshot 2024-06-05 at 18 15 18" src="https://github.com/radixdlt/babylon-wallet-android/assets/164897324/fc6dc988-d551-46fb-be41-6a33b4d24b6d">

Reply:
I think you are only in one place wrong. Unless we are talking/thinking different things. 😅 

> All the computation done upon collecting or in the operators called from the viewModel will still be on main thread if moving the flowOn to the use cases.

If you move the `.flowOn(Dispatchers.Default)` at the end of the `GetSecurityProblemsUseCase` all the computations of this use case will run in the Default Dispatchers.

And then only the `_state.emit(` will run on the Main. 

Do we agree so far?

If yes then would we have a performance impact on rendering this particular screen (Security Center)? Tbh I am not sure about this, but I assume no.

Reply:
my initial idea was to "seal" these usecases by running them on Default dispatcher and then we don't need to care when we use them in VMs or anywhere else

Reply:
That's right! But look what we pass to `_state.emit(` it's a new `SecurityCenterUiState` object that upon instantiation also initializes 3 boolean values, those will also be computed on the main thread and I agree that it won't make any difference here specifically. I was trying to understand what's the issue with having the `flowOn(dispatcher)` inside the `viewModel` in general, because we have many places where a lot more computation is done there.

Reply:
no there's not an issue - forget about my second comment.

my initial thought was this one but I should have expressed it in a better way 😓 

>my initial idea was to "seal" these usecases by running them on Default dispatcher and then we don't need to care when we use them in VMs or anywhere else

Reply:
You're right, having the dispatcher specified at a `UseCase` layer level sounds much better than having it chaotically spread across the whole codebase.